src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp

```diff
@@ -247,9 +247,11 @@ inline oop PSPromotionManager::copy_unmarked_to_survivor_space(oop o,
   // Copy obj
   Copy::aligned_disjoint_words(cast_from_oop<HeapWord*>(o), cast_from_oop<HeapWord*>(new_obj), new_obj_size);
 
-  // Parallel GC claims with a release - so other threads might access this object
-  // after claiming and they should see the "completed" object.
-  ContinuationGCSupport::transform_stack_chunk(new_obj);
+  if (!new_obj->mark().is_marked()) {
```

Comment:
For this check to work correctly, we are assuming that Copy::aligned_disjoint_words respects word level atomicity, even though we are using one of the non-atomic copying functions. That doesn't feel safe.

Reply:
True, it is not exactly safe. I wonder if we can plug this particular leak by doing the 

```
  if (UseCompactObjectHeaders) {
    // The copy above is not atomic. Make sure we have seen the proper mark
    // and re-install it into the copy, so that Klass* is guaranteed to be correct.
    markWord mark = o->mark_acquire();
    if (!mark.is_marked()) {
      new_obj->set_mark(mark);
      ContinuationGCSupport::transform_stack_chunk(new_obj);
    } else {
      // If we copied a mark-word that indicates 'forwarded' state, the object
      // installation would not succeed. We cannot access Klass* anymore either.
      // Skip the transformation.
    }
  } else {
    ContinuationGCSupport::transform_stack_chunk(new_obj);
  }
```

Reply:
The load in mark_acquire can float up above the copying. So I don't think that will work either.

Reply:
Hmm, right. I guess this is not only about atomicity. It's also possible that we see that it's not marked/forwarded, then ignore the transform_stack_chunk() call, which would be wrong.
The problem is that transform_stack_chunk() wants to access the Klass* to check is_stackChunk(). So maybe we need to extract the Klass* from the test_mark and pass it to (a new variant of) ContinuationSupport::transform_stack_chunk() which only uses that class? That should work, right?

Reply:
I don't quite see the problem. So the `mark_acquire` floats before the copy, what does it break? 

If we read the "forwarded" mark early, we know it would fail installation, because there is already a forwarding. No transformation is needed then.

If we read the "not forwarded" mark, then there a chance we are going to install the copy as forwardee, so we need to transform. But for the case of compact object headers, we need to guarantee for `transform_stack_chunk` is seeing a consistent `Klass*` from the mark, which specially atomic read of mark provides. We store that mark in the copy too, thus overwriting any atomicity violations the copy routine might have introduced, and giving `transform_chunk_call` a green light to proceed with transformation. (Aside: This thing could even be just `mark`, because even the relaxed atomic load would suffice.)

IOW, if the copy routine is not atomic, and it is a problem for mark, then we can overwrite mark in the copy with the special atomic read.

Reply:
> Hmm, right. I guess this is not only about atomicity. It's also possible that we see that it's not marked/forwarded, then ignore the transform_stack_chunk() call, which would be wrong.
> The problem is that transform_stack_chunk() wants to access the Klass* to check is_stackChunk(). So maybe we need to extract the Klass* from the test_mark and pass it to (a new variant of) ContinuationSupport::transform_stack_chunk() which only uses that class? That should work, right?

Yes, that would work. Given of course that nobody is reloading the klass somewhere in there.

Reply:
> I don't quite see the problem. So the `mark_acquire` floats before the copy, what does it break?
> 
> If we read the "forwarded" mark early, we know it would fail installation, because there is already a forwarding. No transformation is needed then.
> 
> If we read the "not forwarded" mark, then there a chance we are going to install the copy as forwardee, so we need to transform. But for the case of compact object headers, we need to guarantee for `transform_stack_chunk` is seeing a consistent `Klass*` from the mark, which specially atomic read of mark provides. We store that mark in the copy too, thus overwriting any atomicity violations the copy routine might have introduced, and giving `transform_chunk_call` a green light to proceed with transformation. (Aside: This thing could even be just `mark`, because even the relaxed atomic load would suffice.)
> 
> IOW, if the copy routine is not atomic, and it is a problem for mark, then we can overwrite mark in the copy with the value we got with special atomic load.

Ah. Yes, I think you are right.